tests/KubernetesClient.Tests/Mock/MockKubeApiServer.cs

```diff
@@ -10,26 +11,206 @@
+        public MockKubeApiServer(ITestOutputHelper testOutput, params MockKubeServerFlag[] serverFlags)
+        {
+            _serverFlags = serverFlags;
+            _webHost = BuildWebHost(testOutput, ShouldNext, null, null); // to avoid making ShouldNext static
+            _webHost.Start();
+        }
+
+        public MockKubeApiServer(
+            ITestOutputHelper testOutput,
+            Func<HttpContext, Task<bool>> shouldNext = default,
+            Action<ListenOptions> listenConfigure = default,
             string resp = MockPodResponse)
+        {
+            _webHost = BuildWebHost(testOutput, shouldNext, listenConfigure, resp);
+            _webHost.Start();
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            // Execute if resources have not already been disposed.
+            if (!_disposed)
+            {
+                // If the call is from Dispose, free managed resources.
+                if (disposing)
+                {
+                    _webHost.StopAsync();
+                    _webHost.WaitForShutdown();
+                    _webHost.Dispose();
+                }
+            }
+
+            _disposed = true;
+        }
+
+        private static string BuildWatchEventStreamLine(WatchEventType eventType)
+        {
+            var corev1PodList = JsonConvert.DeserializeObject<V1PodList>(MockPodResponse);
+            return JsonConvert.SerializeObject(
+                new Watcher<V1Pod>.WatchEvent { Type = eventType, Object = corev1PodList.Items.First() },
+                new StringEnumConverter());
+        }
+
+        private async Task WriteStreamLine(HttpContext httpContext, string reponseLine)
+        {
+            const string crlf = "\r\n";
+            await httpContext.Response.WriteAsync(reponseLine.Replace(crlf, "")).ConfigureAwait(false);
+            await httpContext.Response.WriteAsync(crlf).ConfigureAwait(false);
+            await httpContext.Response.Body.FlushAsync().ConfigureAwait(false);
+        }
+
+        private async Task<bool> ShouldNext(HttpContext httpContext)
+        {
+            bool Contains(MockKubeServerFlag flag)
+            {
+                return _serverFlags is { } && _serverFlags.Contains(flag);
+            }
+
+            var isWatch = (httpContext.Request.Query.ContainsKey("watch") && httpContext.Request.Query["watch"] == "true");
+            var returnStatusCode = (Contains(MockKubeServerFlag.Throw500) ? HttpStatusCode.InternalServerError : HttpStatusCode.OK);
+
+            httpContext.Response.StatusCode = (int)returnStatusCode;
+            httpContext.Response.ContentLength = null;
+
+            if (isWatch && !Contains(MockKubeServerFlag.Throw500))
+            {
+                ServerShutdown = new AsyncManualResetEvent();
+                if (Contains(MockKubeServerFlag.AddedPod))
```

Comment:
better style should be
`all flags`.map to `stream line`. foreach `writestreamline`